lib/web/conn_upgrade.go

```diff
@@ -84,6 +95,48 @@ func (h *Handler) connectionUpgrade(w http.ResponseWriter, r *http.Request, p ht
 	return nil, nil
 }
 
+func (h *Handler) upgradeALPNWebSocket(w http.ResponseWriter, r *http.Request, upgradeHandler ConnectionHandler) (interface{}, error) {
+	upgrader := websocket.Upgrader{
+		CheckOrigin: func(r *http.Request) bool { return true },
+		Subprotocols: []string{
+			constants.WebAPIConnUpgradeTypeALPN,
+			constants.WebAPIConnUpgradeTypeALPNPing,
+		},
+	}
+	wsConn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		h.log.WithError(err).Debug("Failed to upgrade weboscket.")
+		return nil, trace.Wrap(err)
+	}
+	defer wsConn.Close()
+
+	logrus.WithField("protocol", wsConn.Subprotocol()).Trace("Received WebSocket upgrade.")
+
+	conn := newWebSocketALPNServerConn(wsConn)
+	ctx, cancel := context.WithCancel(r.Context())
+	defer cancel()
+
+	switch wsConn.Subprotocol() {
+	case constants.WebAPIConnUpgradeTypeALPNPing:
+		// Starts native WebSocket ping for "alpn-ping".
+		go h.startPing(ctx, conn)
+	case constants.WebAPIConnUpgradeTypeALPN:
+		// Nothing to do
+	default:
+		// Just close the connection. Upgrader hijacks the connection so no
+		// point returning an error.
+		h.log.WithField("client-protocols", websocket.Subprotocols(r)).
+			Debug("Unknown or empty WebSocket subprotocol.")
+		return nil, nil
+	}
```

Comment:
Should we add the 	handler for `WebAPIConnUpgradeTypeWebSocket`. I can imagine that the https://github.com/gravitational/teleport/pull/36343/files#diff-ddfc49012fd4800c5b6f235daf4c852668a2053bdcea1910a374286766744a91R230 block can be delated resulting in sending the  `WebAPIConnUpgradeTypeWebSocket`  from the clients.

Reply:
This is the sub-protocol received from `Sec-WebSocket-Protocol`, i reused previous `alpn` and `alpn-ping` as names.